### PR TITLE
Components: Refactor uPlot component logic into hooks and libs

### PR DESCRIPTION
**Pull request by jsnmoon**
Related to #76328.

## Proposed Changes

* Moves the chart resize hook outside of the component file.
* Creates a new hook wrapping the existing logic for creating a scale gradient.
* Creates a new lib for defining a gradient fill which is used for a series' stroke.

## Testing Instructions

* Open the live branch for this PR.
* Navigate to `/stats/day/:siteSlug`.
* Add `?flags=stats/subscribers-section` to the URL to enable the subscriber stats page.
* Ensure that the chart loads as expected.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: packages/components/src/chart-uplot/hooks/use-resize.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import throttle from 'lodash/throttle';
```
Comment (jsnmoon) [inline packages/components/src/chart-uplot/hooks/use-resize.ts:1]: Moved from `index.tsx` without modifications.

File: packages/components/src/chart-uplot/hooks/use-scale-gradient.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+export type ScaleGradientFunction = (
+	u: uPlot,
+	scaleKey: string,
+	ori: number,
+	scaleStops: [ number, string ][],
+	discrete?: boolean
+) => CanvasRenderingContext2D[ 'fillStyle' ];
+
+export default function useScaleGradient( backupColor: string ) {
```
Comment (jsnmoon) [inline packages/components/src/chart-uplot/hooks/use-scale-gradient.ts:12]: Moved from `index.tsx` without modifications to logic. Variable names have been updated for clarity.

File: packages/components/src/chart-uplot/lib/get-gradient-fill.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import type { ScaleGradientFunction } from '../hooks/use-scale-gradient';
+import type uPlot from 'uplot';
+
+export default function getGradientFill(
```
Comment (jsnmoon) [inline packages/components/src/chart-uplot/lib/get-gradient-fill.ts:4]: Extracted from the options object declaration in `index.tsx` without modification.

**Review by grzegorz-cp — APPROVED** [review]
Looks good. Thank you for proposing the new structure.

**Pull request merged**